src/Controller/Component/SecurityComponent.php

```diff
@@ -344,20 +446,99 @@ protected function _validatePost(Controller $controller)
+        $messages = [];
+        $expectedParts = json_decode(urldecode($controller->request->data['_Token']['debug']), true);
+        if (!is_array($expectedParts) || count($expectedParts) !== 3) {
+            return 'Invalid security debug token.';
+        }
+        if ($hashParts[0] !== $expectedParts[0]) {
+            $messages[] = sprintf('URL mismatch in POST data (expected \'%s\' but found \'%s\')', $expectedParts[0], $hashParts[0]);
+        }
+        $expectedFields = $expectedParts[1];
+        $dataFields = unserialize($hashParts[1]);
+        $fieldsMessages = $this->_debugCheckFields(
+            $dataFields,
+            $expectedFields,
+            'Unexpected field \'%s\' in POST data',
+            'Tampered field \'%s\' in POST data (expected value \'%s\' but found \'%s\')',
+            'Missing field \'%s\' in POST data'
+        );
+        $expectedUnlockedFields = Hash::get($expectedParts, 2);
+        $dataUnlockedFields = Hash::get($hashParts, 2) ?: [];
+        if ($dataUnlockedFields) {
+            $dataUnlockedFields = explode('|', $dataUnlockedFields);
+        }
+        $unlockFieldsMessages = $this->_debugCheckFields(
+            $dataUnlockedFields,
+            $expectedUnlockedFields,
+            'Unexpected unlocked field \'%s\' in POST data',
+            null,
+            'Missing unlocked field: \'%s\''
+        );
+
+        $messages = array_merge($messages, $fieldsMessages, $unlockFieldsMessages);
+        return implode(', ', $messages);
+    }
+
+    /**
+     * Iterates data array to check against expected
+     *
+     * @param array $dataFields Fields array, containing the POST data fields
+     * @param array $expectedFields Fields array, containing the expected fields we should have in POST
+     * @param string $intKeyMessage Message string if unexpected found in data fields indexed by int (not protected)
+     * @param string $stringKeyMessage Message string if tampered found in data fields indexed by string (protected)
+     * @param string $missingMessage Message string if missing field
+     * @return array Messages
+     */
+    protected function _debugCheckFields($dataFields, $expectedFields = [], $intKeyMessage = '', $stringKeyMessage = '', $missingMessage = '')
+    {
+        $messages = [];
+        foreach ($dataFields as $key => $value) {
+            if (is_int($key)) {
+                $foundKey = array_search($value, (array)$expectedFields);
+                if ($foundKey === false) {
+                    $messages[] = sprintf($intKeyMessage, $value);
+                } else {
+                    unset($expectedFields[$foundKey]);
+                }
+            } elseif (is_string($key)) {
+                if (isset($expectedFields[$key]) && $value !== $expectedFields[$key]) {
+                    $messages[] = sprintf($stringKeyMessage, $key, $expectedFields[$key], $value);
+                }
+                unset($expectedFields[$key]);
+            }
+        }
```

Comment:
What about errors on deeply nested fields? Comparing the two sets of fields after normalizing with `Hash::flatten()` might be simpler than trying to recurse through 2 trees of data.